pkg/odo/genericclioptions/context.go

```diff
@@ -0,0 +1,204 @@
+	projectFlag := FlagValueIfSet(command, util.ProjectFlagName)
+	if len(projectFlag) > 0 {
+		// if project flag was set, check that the specified project exists and use it
+		_, err := project.Exists(client, projectFlag)
+		util.CheckError(err, "")
+		ns = projectFlag
+	} else {
+		// otherwise use the current project
+		ns = project.GetCurrent(client)
+	}
+	client.Namespace = ns
+
+	// resolve application
+	var app string
+	appFlag := FlagValueIfSet(command, util.ApplicationFlagName)
+	if len(appFlag) > 0 {
+		// if we specified an application via flag, check that it exists and use it
+		_, err := application.Exists(client, appFlag)
+		util.CheckError(err, "")
+		app = appFlag
+	} else {
+		var err error
+		if !createAppIfNeeded {
+			// otherwise get the current app (which might not exist)
+			app, err = application.GetCurrent(ns)
+		} else {
+			// if we asked an app to be created if missing, get the existing one or creating one if needed
+			app, err = application.GetCurrentOrGetCreateSetDefault(client)
+		}
+		util.CheckError(err, "unable to get current application")
+	}
+
+	// create the internal context representation based on calculated values
+	internalCxt := internalCxt{
+		Client:      client,
+		Project:     ns,
+		Application: app,
+	}
+
+	// create a context from the internal representation
+	context := &Context{
+		internalCxt: internalCxt,
+	}
+
+	// resolve component
+	var cmp string
+	cmpFlag := FlagValueIfSet(command, util.ComponentFlagName)
+	if len(cmpFlag) == 0 {
+		// retrieve the current component if it exists if we didn't set the component flag
+		var err error
+		cmp, err = component.GetCurrent(app, ns)
+		util.CheckError(err, "could not get current component")
+	} else {
+		// if flag is set, check that the specified component exists
+		context.existsOrExit(cmpFlag)
+		cmp = cmpFlag
+	}
+
+	// once the component is resolved, add it to the context
+	context.cmp = cmp
+
+	return context
+}
+
+// FlagValueIfSet retrieves the value of the specified flag if it is set for the given command
+func FlagValueIfSet(cmd *cobra.Command, flagName string) string {
+	flag, err := cmd.Flags().GetString(flagName)
+
+	// log the error for debugging purposes though an error should only occur if the flag hadn't been added to the command or
+	// if the specified flag name doesn't match a string flag. This usually can be ignored.
+	ignoreButLog(err)
+	return flag
+}
+
+// Context holds contextual information useful to commands such as correctly configured client, target project and application
+// (based on specified flag values) and provides for a way to retrieve a given component given this context
+type Context struct {
+	internalCxt
+}
+
+// internalCxt holds the actual context values and is not exported so that it cannot be instantiated outside of this package.
+// This ensures that Context objects are always created properly via NewContext factory functions.
+type internalCxt struct {
+	Client      *occlient.Client
+	Project     string
```

Comment:
The client already has `Namespace` why project again?

Also, why is fetching project, app and component all 3 always(for every command) required?
Would fetching them on demand for any command be like it used to be before, better?

Reply:
I Agree its not much of an overhead though :thinking:

Reply:
Regarding `Namespace` vs. `Project`, I used the `odo` terminology as we're not using `Namespace` in the rest of the code. This also matches the names of the flags that influence that state.

Reply:
Hmm so since the Context would now capture the project, can we remove Namespace from within client as it can be ambiguous. We can also take that up later if it makes this a very huge change but presence of 2 similar fields leads to ambiguity right?

Reply:
I agree that we should probably remove it but I think that should be the scope of a different PR. Actually, I think that most functions should take a `Context` as parameter as well where it makes sense.

Reply:
> I agree that we should probably remove it but I think that should be the scope of a different PR. Actually, I think that most functions should take a `Context` as parameter as well where it makes sense.

👍  Definitely, we have to refactor most of the function to take this. But it should be addressed in another PR. Meanwhile, we have to be careful with setting Project/Namespace. @anmolbabu has a good point this duplicity could lead to bugs.
@metacosm Could you please create an issue to make sure all this that we need to do won't get lost?

Reply:
Done, see #968